common/display/displayplanemanager.cpp

```diff
@@ -458,6 +460,39 @@ void DisplayPlaneManager::ValidateCachedLayers(
   *render_layers = needs_gpu_composition;
 }
 
+bool DisplayPlaneManager::ValidateSurfaceDamage(
+    const std::vector<OverlayLayer> *layers,
+    const std::vector<OverlayLayer> *previous_layers) {
+  auto layer_begin = layers->begin();
+  auto layer_end = layers->end();
+  bool validate = true;
+  size_t num_layers = layers->size();
+
+  for (size_t i = 0; i <= num_layers; i++) {
+    const HwcRegion &current_surface_damage = layers->at(i).GetSurfaceDamage();
+
+    if (current_surface_damage.kNumRects == 1) {
+      const HwcRect<int> *rect = current_surface_damage.kRects;
+      if ((rect->top == 0) && (rect->bottom == 0) && (rect->left == 0) &&
+          (rect->right == 0)) {
+        const HwcRect<int> &previous = previous_layers->at(i).GetDisplayFrame();
+        const HwcRect<int> &current = layers->at(i).GetDisplayFrame();
+        if ((previous.left != current.left) || (previous.top != current.top)) {
+          validate = true;
+          break;
+        }
+        validate = false;
+        continue;
+      }
+    } else if (current_surface_damage.kNumRects == 0) {
+      validate = true;
```

Comment:
In cases where we fall back to 3D, we can check https://github.com/01org/IA-Hardware-Composer/blob/master/common/display/displayplanemanager.cpp#L439 here if damage region(rects) of that layer is zero. If region_changed is false and damaged(assuming thats the boolean to track the damage state of layers in this local func) we can assign the buffer associated with last_plane here to the current plane. 

In case we are just scanning out directly need to add surface damage check here https://github.com/01org/IA-Hardware-Composer/blob/master/common/display/displayplanemanager.cpp#L452 and swap previous layer buffer to current layer. This way state is always intact at layer level making it easy in future to track damage for per layer.